### PR TITLE
C++/C#: Add `AliasedUse` instruction to all functions

### PR DESCRIPTION
**Pull request by dave-bartolomeo**
This new instruction is the dual of the existing `AliasedDefinition` instruction. Whereas that instruction defines the contents of aliased memory before the function was called, `AliasedUse` represents the potential use of all aliased memory after the function returns. This ensures that writes to aliased memory do not appear "dead", even if there are no further reads from aliased memory within the function itself.

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/Instruction.qll [hunk 1]
```diff
@@ -1423,6 +1425,13 @@ class AliasedDefinitionInstruction extends Instruction {
   final override MemoryAccessKind getResultMemoryAccess() { result instanceof EscapedMemoryAccess }
 }
 
+/**
+ * An instruction that consumes all escaped memory on exit from the function.
+ */
+class AliasedUseInstruction extends Instruction {
+  AliasedUseInstruction() { getOpcode() instanceof Opcode::AliasedUse }
+}
+
```
[the remark this answers is not in the document]
Comment (dbartol) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/Instruction.qll]: The short answer is that nobody should look at `UnmodeledUseInstruction`. That exists solely as a way to ensure that instructions whose results aren't modeled in the current flavor of SSA don't appear dead. Really, the answer there is to not try to do dead code analysis except on full aliased SSA, or to have it take unmodeled memory into account.

I believe that @rdmarsh2 is adding a sanity check to ensure that all memory accesses are modeled in aliased SSA. Once we're confident of that, I think we should remove the `Unmodeled[Use|Def]` instructions from aliased SSA IR completely.

A more specific answer is that `Aliased[Def|Use]` represents memory that is modeled in SSA, but is defined before the current function is entered or used after the current function exits. `Unmodeled[Def|Use]` is for memory that the current flavor of SSA is not able to model at all, so we don't know where it comes from or where it gets used.

**Comment by dbartol** [comment]
@jbj Please take a look at the commit that I added since your review. This commit contains changes that helped improve the results on the insecure `memset` query pretty dramatically.

**Comment by jbj** [comment]
The diff of `aliased_ssa_ir.expected` in 56cbd0c152dade9273b8522775bc9c035825c8e7 shows that the commit had the intended effect on `ModeledCallTarget`, but I'm less sure about the other functions. For example, the operand of `AliasedUse` in `MustExactlyOverlapEscaped` changed just slightly, but I'd think it should change all the way back to the `AliasedDefinition` instruction since neither `a` nor `b` outlive the stack frame.

**Comment by dbartol** [comment]
In `MustExactlyOverlapEscaped`, neither of the two local variables outlive the stack frame, but the `CallSideEffect` instruction that defines `m0_14` is accessing all aliased memory, not just local variables.

**Pull request merged**